live-build/misc/migration-scripts/dx_delete

```diff
@@ -39,10 +39,94 @@ function die() {
 # names are randomly generated.
 #
 
-linux_dataset="rpool/ROOT"
-if ! zfs list ${linux_dataset} &>/dev/null; then
-	echo "Linux dataset '${linux_dataset}' is not installed."
+LX_DATASET="rpool/ROOT"
+if ! zfs list ${LX_DATASET} &>/dev/null; then
+	echo "Linux dataset '${LX_DATASET}' is not installed."
 	exit 0
 fi
 
-zfs destroy -r ${linux_dataset} || die "Failed to destroy Linux dataset '${linux_dataset}'"
+LX_RDS_PARENT=$(zfs list -o name -H -d 1 "$LX_DATASET" | tail -n 1)
+[[ -n $LX_RDS_PARENT ]] || die "could not find Linux RDS parent dataset"
+LX_CONTAINER="${LX_RDS_PARENT##*/}"
+LX_RDS_TMP_ROOT_MOUNT="/tmp/$LX_CONTAINER/root"
+
+MDS_SNAPNAME="MDS-CLONE-upgradeverify"
+MDS_CLONE=domain0/$MDS_SNAPNAME
+SVC=svc:/system/delphix/postgres:$MDS_SNAPNAME
+PG_DATA=/$MDS_CLONE/db
+
+function dx_execute_cleanup() {
+	#
+	# Undo any existing default timeout commands in the bootloader's menu.
+	#
+	sed -i '/menu_timeout_command/d' /boot/menu.rc.local
+	[[ "$(grep -cF 'menu_timeout_command' /boot/menu.rc.local)" -eq 0 ]] ||
+		die "failed to cleanup previous default command"
+
+	#
+	# We could also delete /var/delphix/migration here but we decide
+	# to leave it for debugging purposes. Re-executing dx_execute always
+	# deletes and recreates the directory so we'll always have the logs
+	# of the latest run.
+	#
+}
+
+function dx_apply_cleanup() {
+	#
+	# Remove new bootloader entry
+	#
+	sed -i '/set lxcmdline/d' /boot/menu.rc.local
+	[[ "$(grep -cF 'set lxcmdline' /boot/menu.rc.local)" -eq 0 ]] ||
+		die "failed to cleanup lxcmdline from previous run"
```

Comment:
This will only remove the command executed by the new 8th option that we added to the bootloader (i.e. the option to boot linux). The actual option will not be gone from the menu, so I think we probably want to add the following lines too:
```
sed -i '/set mainmenu_caption\[8\]/d' /boot/menu.rc.local
[[ "$(grep -cF 'mainmenu_caption[8]' /boot/menu.rc.local)" -eq 0 ]] ||
	die "failed to cleanup mainmenu_caption from previous run"
sed -i '/set mainansi_caption\[8\]/d' /boot/menu.rc.local
[[ "$(grep -cF 'mainansi_caption[8]' /boot/menu.rc.local)" -eq 0 ]] ||
	die "failed to cleanup mainansi_caption from previous run"
sed -i '/set mainmenu_keycode\[8\]/d' /boot/menu.rc.local
[[ "$(grep -cF 'mainmenu_keycode[8]' /boot/menu.rc.local)" -eq 0 ]] ||
	die "failed to cleanup mainmenu_keycode from previous run"
sed -i '/set mainmenu_command\[8\]/d' /boot/menu.rc.local
[[ "$(grep -cF 'mainmenu_command[8]' /boot/menu.rc.local)" -eq 0 ]] ||
	die "failed to cleanup mainmenu_command from previous run"
```

In addition we leave some leftover state in the `/boot` directory of illumos that could be cleaned:
```
rm -f /boot/vmlinuz-* /boot/initrd.img-* ||
	die "failed to destroy previously copied Linux kernel data"
```

Reply:
Another approach to consider would be to save the original file so that it can be copied back on cleanup from failure.

Reply:
Yeah, I could see that being pretty clean. At the beginning of `dx_apply` we could save the file in `/boot/menu.rc.local.copy` and the restore it in `dx_delete`. What do you think of this approach @sdimitro ?

Reply:
Yeah that makes sense to me!